ambry-protocol/src/main/java/com/github/ambry/protocol/AmbryRequests.java

```diff
@@ -110,46 +112,69 @@ public AmbryRequests(StoreManager storeManager, RequestResponseChannel requestRe
+      }
       switch (type) {
         case PutRequest:
-          handlePutRequest(request);
+          handlePutRequest(networkRequest);
           break;
         case GetRequest:
-          handleGetRequest(request);
+          handleGetRequest(networkRequest);
           break;
         case DeleteRequest:
-          handleDeleteRequest(request);
+          handleDeleteRequest(networkRequest);
           break;
         case TtlUpdateRequest:
-          handleTtlUpdateRequest(request);
+          handleTtlUpdateRequest(networkRequest);
           break;
         case ReplicaMetadataRequest:
-          handleReplicaMetadataRequest(request);
+          handleReplicaMetadataRequest(networkRequest);
           break;
         case AdminRequest:
-          handleAdminRequest(request);
+          handleAdminRequest(networkRequest);
           break;
         case UndeleteRequest:
-          handleUndeleteRequest(request);
+          handleUndeleteRequest(networkRequest);
           break;
         default:
           throw new UnsupportedOperationException("Request type not supported");
       }
     } catch (Exception e) {
-      logger.error("Error while handling request {} closing connection", request, e);
-      requestResponseChannel.closeConnection(request);
+      logger.error("Error while handling request {} closing connection", networkRequest, e);
+      requestResponseChannel.closeConnection(networkRequest);
     }
   }
 
   @Override
   public void handlePutRequest(NetworkRequest request) throws IOException, InterruptedException {
-    InputStream is = request.getInputStream();
-    DataInputStream dis = is instanceof DataInputStream ? (DataInputStream) is : new DataInputStream(is);
-    PutRequest receivedRequest = PutRequest.readFrom(dis, clusterMap);
+    PutRequest receivedRequest;
+    if (request instanceof LocalChannelRequest) {
+      PutRequest sentRequest = (PutRequest) ((LocalChannelRequest) request).getRequestInfo().getRequest();
+      // Create a new PutRequest object to represent the received Put request since the blob content is carried and
+      // accessed via {@code PutRequest#blobStream} at the receiver side where as it is carried via {@code PutRequest#blob}
+      // at the sender side. Also, using null crc value since the request is sent and received in the same process via
+      // local queues.
+      // Since we are creating ByteBufInputStream from the bytebuf object in sent request, setting releaseOnClose to
+      // true to release it once the stream is closed.
+      receivedRequest =
+          new PutRequest(sentRequest.getCorrelationId(), sentRequest.getClientId(), sentRequest.getBlobId(),
+              sentRequest.getBlobProperties(), sentRequest.getUsermetadata(), sentRequest.getBlobSize(),
+              sentRequest.getBlobType(), sentRequest.getBlobEncryptionKey(),
+              new ByteBufInputStream(sentRequest.getBlob(), true), null);
```

Comment:
What was the reason for not carrying over the CRC? The CRC is also used for some use cases like exact duplicate put entry detection (although some of these may not apply often for cloud use cases since there is no intracolo replication).